### PR TITLE
🚇 Public App File Index Handling

### PR DESCRIPTION
**Pull request by dmiller15**
## Description

This PR addresses the changes necessary to make the workflow public app ready. Specifically these changes address the inability of having suggested secondaryFiles when running a public app.

The changes are made in the following ways:
- Added several tools to optionally perfom indexing (tabix, picard, bwa, samtools)
- Added a subworkflow that indexes the reference fasta and bundles those indexes as secondaryFiles
- Added a subworkflow that can untar a specifically designed tar that contains the reference fasta and its indexes
- Added additional tar tools that provide other tar utilities (listing and creation)

Changes to the workflow structure
- Added a step for dbSNP VCF index creation in the gVCF creation subworkflow
- Added input hook for dbSNP index to the gVCF creation subworkflow and propagated it out to the main workflow
- Input reference is changed from a file with secondaryFiles to a file (documentation and id indicate we are seeking a TAR for user clarity)
- Added steps in the main workflow to untar the reference tar, index as needed (in reality all of these steps will be skipped because the tar MUST have the files in it or the untarring process will fail). The result will be an indexed_reference_fasta which is used in all the same places the old one was
- Added a scattered step to index the items in the knownsites file list.

Resolves https://github.com/d3b-center/bixu-tracker/issues/661

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Tool passes cwltool validation
- [x] Cavatica test: https://cavatica.sbgenomics.com/u/kfdrc-harmonization/kf-reference-pipeline/tasks/08705e71-0b4a-4811-9bbd-43997a706cf5/
- [x] Cavatica test of new knownsite index change: https://cavatica.sbgenomics.com/u/kfdrc-harmonization/kf-reference-pipeline/tasks/9442b46d-ca83-402d-90a3-7a672f41dfc2/

**Test Configuration**:
* Environment: cwltool 3.0.20200324120055, Cavatica
* Test files: See Cavatica test

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -89,23 +89,46 @@ also be provided: `wxs_bait_interval_list` and `wxs_target_interval_list` for Hs
    that the first file in `input_pe_reads_list` is run with the first file in `input_pe_mates_list`
    and the first string in `input_pe_rgs_list`. This also means these arrays must be the same
    length or the workflow will fail.
-1. Must have these associated indexes:
-    - knownsite vcfs: Each file requires a `'.tbi'` index
-    - reference fasta: BWA, picard, and samtools indexes (`'.64.amb', '.64.ann', '.64.bwt',
-        '.64.pac', '.64.sa', '.64.alt', '^.dict', '.fai'`)
+1. The expected input for the reference_tar is a tar file containing the reference fasta along with its indexes.
```
Comment (migbro) [inline readme.md:92]: Hmmm if a user wanted to use hg19, or if it wasn't the `.64` extension style, would this fail?

File: subworkflows/untar_reference_tar.cwl [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    run: ../tools/untar.cwl
+    in:
+      tarfile: reference_tar
+      extractfile: { valueFrom: "Homo_sapiens_assembly38.fasta" }
+    out: [output]
+  untar_fai:
+    run: ../tools/untar.cwl
+    in:
+      tarfile: reference_tar
+      extractfile: { valueFrom: "Homo_sapiens_assembly38.fasta.fai" }
```
Comment (migbro) [inline subworkflows/untar_reference_tar.cwl:34]: Hmmm, yeah, this seems too strict, maybe a workaround could be to use the basename of the input fasta to seed the value and/or have the user specify whether it's `-6` style naming or not.  Hopefully that wouldn't be too difficult while being flexible.  Makes sense?

File: readme.md (hunk 1 above)
Comment (dmiller15) [reply readme.md:92]: Yes. The untarring process relies on pretty exact contents for the inputs.

File: workflows/kfdrc_alignment_wf_cyoa.cwl [hunk 3]
```diff
@@ -66,12 +76,13 @@ inputs:
   input_pe_rgs_list: { type: 'string[]?', doc: "List of RG strings to use in PE processing" }
   input_se_reads_list: { type: 'File[]?', doc: "List of input singlie end fastq reads" }
   input_se_rgs_list: { type: 'string[]?', doc: "List of RG strings to use in SE processing" }
-  indexed_reference_fasta: { type: File, secondaryFiles: ['.64.amb', '.64.ann', '.64.bwt', '.64.pac', '.64.sa', '.64.alt', '^.dict', '.fai'], doc: "Reference fasta with BWA and samtool indexes" }
+  reference_tar: { type: File, doc: "Tar file containing reference fasta and its associated indexes (samtools, bwa, and picard)" }
```
Comment (migbro) [inline workflows/kfdrc_alignment_wf_cyoa.cwl:79]: Would it be too difficult/messy to have a user either give the reference fasta, which activates sort of a secondary file mode, or a tar ball? Am I contradicting myself form previous requests/conversations?

File: readme.md [hunk 4]
```diff
@@ -89,23 +89,51 @@ also be provided: `wxs_bait_interval_list` and `wxs_target_interval_list` for Hs
    that the first file in `input_pe_reads_list` is run with the first file in `input_pe_mates_list`
    and the first string in `input_pe_rgs_list`. This also means these arrays must be the same
    length or the workflow will fail.
-1. Must have these associated indexes:
-    - knownsite vcfs: Each file requires a `'.tbi'` index
-    - reference fasta: BWA, picard, and samtools indexes (`'.64.amb', '.64.ann', '.64.bwt',
-        '.64.pac', '.64.sa', '.64.alt', '^.dict', '.fai'`)
+1. The input for the reference_tar must be a tar file containing the reference fasta along with its indexes.
+   The required indexes are `[.ann,.amb,.bwt,.pac.sa,.dict,.fai]` and are generated by bwa, picard, and samtools.
```
Comment (migbro) [inline readme.md:93]: mod to say `.64`

File: subworkflows/kfdrc_process_bamlist.cwl [hunk 5]
```diff
@@ -9,7 +9,7 @@ inputs:
   input_bam_list: File[]
   indexed_reference_fasta:
     type: File
-    secondaryFiles: ['.64.amb', '.64.ann', '.64.bwt', '.64.pac', '.64.sa', '.64.alt', '^.dict']
+    secondaryFiles: ['.amb', '.ann', '.bwt', '.pac', '.sa', '^.dict', '.fai']
```
Comment (migbro) [inline subworkflows/kfdrc_process_bamlist.cwl:12]: make `.64`

File: subworkflows/kfdrc_process_pe_readslist2.cwl [hunk 6]
```diff
@@ -12,7 +12,7 @@ inputs:
   conditional_run: int
   indexed_reference_fasta:
     type: File
-    secondaryFiles: ['.64.amb', '.64.ann', '.64.bwt', '.64.pac', '.64.sa', '.64.alt', '^.dict']
+    secondaryFiles: ['.amb', '.ann', '.bwt', '.pac', '.sa', '^.dict', '.fai']
```
Comment (migbro) [inline subworkflows/kfdrc_process_pe_readslist2.cwl:15]: make `.64`

File: subworkflows/kfdrc_process_se_readslist2.cwl [hunk 7]
```diff
@@ -11,7 +11,7 @@ inputs:
   conditional_run: int
   indexed_reference_fasta:
     type: File
-    secondaryFiles: ['.64.amb', '.64.ann', '.64.bwt', '.64.pac', '.64.sa', '.64.alt', '^.dict']
+    secondaryFiles: ['.amb', '.ann', '.bwt', '.pac', '.sa', '^.dict', '.fai']
```
Comment (migbro) [inline subworkflows/kfdrc_process_se_readslist2.cwl:14]: make `.64`

File: workflows/kfdrc_alignment_wf_cyoa.cwl [hunk 8]
```diff
@@ -35,22 +35,37 @@ doc: |-
       that the first file in `input_pe_reads_list` is run with the first file in `input_pe_mates_list`
       and the first string in `input_pe_rgs_list`. This also means these arrays must be the same
       length or the workflow will fail.
-   1. Must have these associated indexes:
-      - knownsite vcfs: Each file requires a '.tbi' index
-      - reference fasta: BWA and samtools indexes ('.64.amb', '.64.ann', '.64.bwt',
-          '.64.pac', '.64.sa', '.64.alt', '^.dict', '.fai')
+   1. The input for the reference_tar must be a tar file containing the reference fasta along with its indexes.
+      The required indexes are `[.ann,.amb,.bwt,.pac.sa,.dict,.fai]` and are generated by bwa, picard, and samtools.
```
Comment (migbro) [inline workflows/kfdrc_alignment_wf_cyoa.cwl:39]: revert to `.64`

**Comment by migbro** [comment]
If you look at the log in the test run, the command just seems to untar the whole thing wach time: `tar xf /sbgenomics/Projects/0154a615-2b20-4c32-9c76-141b5eebdde1/Homo_sapiens_assembly38.tgz`
from https://cavatica.sbgenomics.com/u/kfdrc-harmonization/kf-reference-pipeline/tasks/54a2b4b2-b928-4f13-96bc-31ffc3530a7f/stats/
I thought it was supposed to grab a specific file?

**Comment by dmiller15** [comment]
I changed the way the tar process works. The untar process was only working when I used exact file names so I just decided to untar everything and sort it out with the glob.

**Pull request merged**